### PR TITLE
Update calling-c-and-fortran docs

### PR DESCRIPTION
**Pull request by kmsquire**
This is an updated version of #10799.  It's almost entirely formatting and wording updates.  

Comments welcome, although I probably won't be able to spend much time on this.  

Maintainers should feel free to push edits.

### CODE REVIEW

**Comment by Gnimuc** [comment]
Maybe we can also remove some misleading docs in this PR as per discussed in https://github.com/JuliaInterop/Clang.jl/issues/120#issuecomment-478370505 ?

**Comment by Gnimuc** [comment]
> Note that for this code to work correctly, result_array **must** be declared to be of type Ref{Cdouble} and not Ptr{Cdouble}. The memory is managed by Julia and **the Ref signature alerts Julia's garbage collector** to keep managing the memory for result_array while the ccall executes.

I suggest to rephrase or remove the above misleading statements. It just throw some "best practice" rules and doesn't explain [the mechanism under the hood](https://github.com/JuliaInterop/Clang.jl/issues/120#issuecomment-478334164).

> result_array **must** be declared to be of type Ref{Cdouble} and not Ptr{Cdouble}.

This is not right. `Ptr{Cdouble}` should also work safely as long as `Base.cconvert(Ptr{Cdouble}, result_array)` doesn't violate the mechanism. Currently, it calls `Base.cconvert(::Type{<:Ptr}, x) = x` where the return value `x` is guaranteed to be alive (GC-preserved) until the end of `ccall`. 

> **the Ref signature alerts Julia's garbage collector** to keep managing the memory for result_array while the ccall executes.

Again, this is not how it works. `Ref` is not being treated specially, instead, there are some default `cconvert`&`unsafe_convert` methods pre-defined in Julia base, which ensures its GC safety.

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 1]
```diff
@@ -259,26 +260,16 @@ to the specified type. For example, the following call:
...
 ```
 
-will behave as if the following were written:
+will behave as if the following were written[^3]:
 
 ```julia
 ccall((:foo, "libfoo"), Cvoid, (Int32, Float64),
-      Base.unsafe_convert(Int32, Base.cconvert(Int32, x)),
-      Base.unsafe_convert(Float64, Base.cconvert(Float64, y)))
+      Base.convert(Int32, x), Base.convert(Float64, y))
```
Comment (fredrikekre) [inline doc/src/manual/calling-c-and-fortran-code.md]: Why change this and remove the text below? This is not how it works and gives the wrong impression.

Comment (kmsquire) [reply doc/src/manual/calling-c-and-fortran-code.md]: The text below is preserved as a footnote.  I wrote it this way because I felt that, for most users, `cconvert` followed by `unsafe_convert` was more detail than they would want to know, and that this gets the point across (i.e., that the passed values are converted to the appropriate type).  That said, I'm willing to change it back if you don't think the footnote is enough or appropriate.

**Comment by kmsquire** [comment]
@Gnimuc I'll try to address your comments today.

> BTW, I think we'd better mention more use cases for `GC.@preserve` in the doc.

I don't have enough grasp of this to add anything useful.  
If you'd like to, why don't you either make a PR against my branch [here](https://github.com/kmsquire/julia/tree/kms/c_fortran_doc_updates) (I'll preserve your commits), or just make a new PR (probably after this is merged).

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 1 above)
Comment (fredrikekre) [reply doc/src/manual/calling-c-and-fortran-code.md]: I mean, this is the docs for how to `ccall` so why should I have to scroll to a footnote in order to see how `ccall` works?

Comment (yuyichao) [reply doc/src/manual/calling-c-and-fortran-code.md]: >  for most users, cconvert followed by unsafe_convert was more detail than they would want to know, and that this gets the point across (i.e., that the passed values are converted to the appropriate type).

For any user, they should either not care about anything about conversion for `ccall` arguments at all, or if they do, they must not related it to `convert`. This change gets the wrong point across. I've recently seen people still thinking `convert` is the method to use for this and it's just wrong.

Bottom line is that I don't really mind if this is said in the footnote or not, but you must not mention `convert`.

**Comment by kmsquire** [comment]
Okay, I've "addressed" @Gnimuc's comments mostly by deleting the offending section.  If anyone has some better wording that could be put in instead, please speak up!

@Gnimuc @fredrikekre @yuyichao If you have time, can you take a glance through the document as it stands right now and let me know if there' s anything else that should be reworded or removed?  Thanks.

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 2]
```diff
@@ -804,13 +778,9 @@ end
...
+that `p` points to may be managed by Julia or by C. A pointer to memory allocated by C should generally
 be of type `Ptr{gsl_permutation}`, but it is convertible using [`Base.cconvert`](@ref) and therefore
-can be used in the same (covariant) context of the input argument to a [`ccall`](@ref). A pointer
-to memory allocated by Julia must be of type `Ref{gsl_permutation}`, to ensure that the memory
-address pointed to is valid and that Julia's garbage collector manages the chunk of memory pointed
-to correctly. Therefore, the `Ref{gsl_permutation}` declaration allows pointers managed by C or
-Julia to be used.
+can be used in the same (covariant) context of the input argument to a [`ccall`](@ref).
 
 If the C wrapper never expects the user to pass pointers to memory managed by Julia, then using
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
Now if you look closely enough at this example, you may notice that it is incorrect, given our explanation above of preferred declaration types. Do you see it? The function we are calling is going to free the memory. This type of operation cannot be given a Julia object (it will crash or cause memory corruption). Therefore, it may be preferable to declare the `p` type as `Ptr{gsl_permutation }`, to make it harder for the user to mistakenly pass another sort of object there than one obtained via `gsl_permutation_alloc`.

If the C wrapper never expects the user to pass pointers to memory managed by Julia, then using
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 3]
```diff
@@ -804,13 +778,9 @@ end
 ```
 
 Here, the input `p` is declared to be of type `Ref{gsl_permutation}`, meaning that the memory
-that `p` points to may be managed by Julia or by C. A pointer to memory allocated by C should
+that `p` points to may be managed by Julia or by C. A pointer to memory allocated by C should generally
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
that `p` points to may be managed by Julia or by C. A pointer to memory allocated by C should
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 4]
```diff
@@ -804,13 +778,9 @@ end
...
 be of type `Ptr{gsl_permutation}`, but it is convertible using [`Base.cconvert`](@ref) and therefore
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
be of type `Ptr{gsl_permutation}`. But `Ptr` is a subtype of `Ref`, and therefore `Ref` be used in any covariant (argument) context.
The `Ref` supertype will accept a `Ptr`, and treats it the same as if it had been restricted to just accept `Ptr`.
Additionally, this makes it possible to pass a handle to memory that has been allocated by Julia. 
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 5]
```diff
@@ -804,13 +778,9 @@ end
...
-can be used in the same (covariant) context of the input argument to a [`ccall`](@ref). A pointer
-to memory allocated by Julia must be of type `Ref{gsl_permutation}`, to ensure that the memory
-address pointed to is valid and that Julia's garbage collector manages the chunk of memory pointed
-to correctly. Therefore, the `Ref{gsl_permutation}` declaration allows pointers managed by C or
-Julia to be used.
+can be used in the same (covariant) context of the input argument to a [`ccall`](@ref).
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 6]
```diff
@@ -841,19 +811,29 @@ end
...
 the Julia pointer to a Julia array data structure into a form understandable by C.
 
-Note that for this code to work correctly, `result_array` must be declared to be of type `Ref{Cdouble}`
-and not `Ptr{Cdouble}`. The memory is managed by Julia and the `Ref` signature alerts Julia's
-garbage collector to keep managing the memory for `result_array` while the [`ccall`](@ref) executes.
-If `Ptr{Cdouble}` were used instead, the [`ccall`](@ref) may still work, but Julia's garbage
-collector would not be aware that the memory declared for `result_array` is being used by the
-external C function. As a result, the code may produce a memory leak if `result_array` never gets
-freed by the garbage collector, or if the garbage collector prematurely frees `result_array`,
-the C function may end up throwing an invalid memory access exception.
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]: 👍

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 7]
```diff
@@ -841,19 +811,29 @@ end
...
+## Fortran Wrapper Example
+
+The following example computes a dot product using a Fortran BLAS function.  Note the use
+of `Ref` and `Ptr`.
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
On every argument type, we specify `Ref` or `Ptr`. This mangling convention may be specific to your fortran compiler and operating system, and is likely undocumented. However, wrapping each in a `Ref` (or `Ptr`, where equivalent) is a frequent requirement of Fortran compiler implementations:
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 8]
```diff
@@ -841,19 +811,29 @@ end
...
-and not `Ptr{Cdouble}`. The memory is managed by Julia and the `Ref` signature alerts Julia's
-garbage collector to keep managing the memory for `result_array` while the [`ccall`](@ref) executes.
-If `Ptr{Cdouble}` were used instead, the [`ccall`](@ref) may still work, but Julia's garbage
-collector would not be aware that the memory declared for `result_array` is being used by the
-external C function. As a result, the code may produce a memory leak if `result_array` never gets
-freed by the garbage collector, or if the garbage collector prematurely frees `result_array`,
-the C function may end up throwing an invalid memory access exception.
+## Fortran Wrapper Example
+
+The following example computes a dot product using a Fortran BLAS function.  Note the use
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
The following example utilizes ccall to call a function in a common Fortran library (libBLAS) to computes a dot product. Notice that the argument mapping is a bit different here than above, as we need to map from Julia to Fortran.
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 9]
```diff
@@ -1079,3 +1059,14 @@ For more details on how to pass callbacks to C libraries, see this [blog post](h
...
 For direct C++ interfacing, see the [Cxx](https://github.com/Keno/Cxx.jl) package. For tools to create C++
 bindings, see the [CxxWrap](https://github.com/JuliaInterop/CxxWrap.jl) package.
+
+
+
+[^1]: Non-library function calls in both C and Julia can be inlined and thus may have
+    even less overhead than calls to shared library functions. When both libraries and executables are
+    generated by LLVM, it is possible to perform whole-program optimizations that can even optimize
+    across this boundary, but Julia does not yet support that. In the future, however, it may do so,
+    yielding even greater performance gains.
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 10]
```diff
@@ -1079,3 +1059,14 @@ For more details on how to pass callbacks to C libraries, see this [blog post](h
 
 For direct C++ interfacing, see the [Cxx](https://github.com/Keno/Cxx.jl) package. For tools to create C++
 bindings, see the [CxxWrap](https://github.com/JuliaInterop/CxxWrap.jl) package.
+
+
+
+[^1]: Non-library function calls in both C and Julia can be inlined and thus may have
+    even less overhead than calls to shared library functions. When both libraries and executables are
+    generated by LLVM, it is possible to perform whole-program optimizations that can even optimize
+    across this boundary, but Julia does not yet support that. In the future, however, it may do so,
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 11]
```diff
@@ -1079,3 +1059,14 @@ For more details on how to pass callbacks to C libraries, see this [blog post](h
 
 For direct C++ interfacing, see the [Cxx](https://github.com/Keno/Cxx.jl) package. For tools to create C++
 bindings, see the [CxxWrap](https://github.com/JuliaInterop/CxxWrap.jl) package.
+
+
+
+[^1]: Non-library function calls in both C and Julia can be inlined and thus may have
+    even less overhead than calls to shared library functions. When both libraries and executables are
+    generated by LLVM, it is possible to perform whole-program optimizations that can even optimize
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
```

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 12]
```diff
@@ -1079,3 +1059,14 @@ For more details on how to pass callbacks to C libraries, see this [blog post](h
 
 For direct C++ interfacing, see the [Cxx](https://github.com/Keno/Cxx.jl) package. For tools to create C++
 bindings, see the [CxxWrap](https://github.com/JuliaInterop/CxxWrap.jl) package.
+
+
+
+[^1]: Non-library function calls in both C and Julia can be inlined and thus may have
+    even less overhead than calls to shared library functions. When both libraries and executables are
```
Comment (vtjnash) [inline doc/src/manual/calling-c-and-fortran-code.md]:
```suggestion
    even less overhead than calls to shared library functions.
    The point above is that the cost of actually doing foreign function call is about the same as doing a call in either native language.
```

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 11 above)
Comment (Gnimuc) [reply doc/src/manual/calling-c-and-fortran-code.md]: just out of curiosity, is this impossible for now?

Comment (vtjnash) [reply doc/src/manual/calling-c-and-fortran-code.md]: no, some packages do it, it's just largely nonsense and irrelevant for the manual

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 13]
```diff
@@ -164,20 +163,20 @@ typedef returntype (*functiontype)(argumenttype, ...)
 ```
 
 The macro [`@cfunction`](@ref) generates the C-compatible function pointer for a call to a
-Julia function. Arguments to [`@cfunction`](@ref) are as follows:
+Julia function. The arguments to [`@cfunction`](@ref) are:
 
 1. A Julia Function
-2. Return type
+2. The return type
```
[the remark this answers is not in the document]
Comment (kmsquire) [reply doc/src/manual/calling-c-and-fortran-code.md]: That depends on how you interpret it.  If you think about the return type as belonging to the function, then I think it should be `the function's return type` (which I shortened to `the return type` here), since the function was introduced above.  If it's an arbitrary return type, then `a return type` would be better.

I'll change it to `the function's return type`, since that's clearer.

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 8 above)
Comment (kmsquire) [reply doc/src/manual/calling-c-and-fortran-code.md]: Here too: will clean up line length later.

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 4 above)
Comment (vtjnash) [reply doc/src/manual/calling-c-and-fortran-code.md]: no, you can never use `pointer`. the rest sounds ok

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 14]
```diff
@@ -485,24 +486,27 @@ are not possible in the translation to Julia.
 
 Packed structs and union declarations are not supported by Julia.
 
-You can get a near approximation of a `union` if you know, a priori, the field that will have
+You can get an approximation of a `union` if you know, a priori, the field that will have
 the greatest size (potentially including padding). When translating your fields to Julia, declare
 the Julia field to be only of that type.
 
-Arrays of parameters can be expressed with `NTuple`:
+Arrays of parameters can be expressed with `NTuple`.  For example, this struct in C,
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply doc/src/manual/calling-c-and-fortran-code.md]: As written, it seems to need a colon, but that's suboptimal because it doesn't match the second clause, which has a comma. Fixing the sentence to have parallel grammar structure resolves the  punctuation issue.

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (vtjnash) [reply doc/src/manual/calling-c-and-fortran-code.md]: OK, I guess I was reading it as "A call's return type"

File: doc/src/manual/calling-c-and-fortran-code.md (hunk 4 above)
Comment (kmsquire) [reply doc/src/manual/calling-c-and-fortran-code.md]: Okay, would I just pass `a` in that second example then?

File: doc/src/manual/calling-c-and-fortran-code.md [hunk 15]
```diff
@@ -164,20 +163,20 @@ typedef returntype (*functiontype)(argumenttype, ...)
...
-Like ccall, all of these arguments will be evaluated at compile-time, when the containing method is defined.
+As with `ccall`, all of these arguments will be evaluated at compile-time when the containing method is defined.
 
 Currently, only the platform-default C calling convention is supported. This means that
 `@cfunction`-generated pointers cannot be used in calls where WINAPI expects `stdcall`
 function on 32-bit windows, but can be used on WIN64 (where `stdcall` is unified with the
 C calling convention).
 
-A classic example is the standard C library `qsort` function, declared as:
+As a classic example, consider the standard C library `qsort` function, declared as:
```
[the remark this answers is not in the document]
Comment (kmsquire) [reply doc/src/manual/calling-c-and-fortran-code.md]: Reverted.

**Comment by kmsquire** [comment]
Updated according to the latest review.  @vtjnash (or anyone else), would you like to take another quick scan?

The test failure is (probably) #31712.
